File: examples/bun/example01_load_version_logs.ts
import {
  DXB_BACKEND_AUTO,
  DXB_FEATURE_DX12,
  DXB_LOG_LEVEL_NAMES,
  DXBridgeLibrary,
  getStringArg,
  parseArgs,
  parseBackend,
  parseVersion,
} from "./dxbridge";

function printUsage(): void {
  console.log("Usage: bun run examples/bun/example01_load_version_logs.ts [--dll PATH] [--backend auto|dx11|dx12]");
}

function main(): void {
  const args = parseArgs(Bun.argv.slice(2));
  if (args.values.help === true) {
    printUsage();
    return;
  }

  const backend = parseBackend(getStringArg(args, "backend", "auto"));
  const dllPath = getStringArg(args, "dll");
  const dxbridge = new DXBridgeLibrary(dllPath || undefined);

  console.log(`Loaded: ${dxbridge.path}`);
  dxbridge.setLogCallback((level, message) => {
    const label = DXB_LOG_LEVEL_NAMES[level] ?? `LEVEL_${level}`;
    console.log(`[log:${label}] ${message}`);
  });

  try {
    const version = dxbridge.getVersion();
    console.log(`DXBridge version: ${parseVersion(version)} (0x${version.toString(16).toUpperCase().padStart(8, "0")})`);

    dxbridge.init(Number.isFinite(backend) ? backend : DXB_BACKEND_AUTO);
    console.log(`DX12 feature flag reported by active backend: ${dxbridge.supportsFeature(DXB_FEATURE_DX12)}`);
    console.log("Initialization and shutdown completed successfully.");
    dxbridge.shutdown();
  } finally {
    dxbridge.close();
  }
}

main();

File: examples/python/example_01_load_version_logs.py
from __future__ import annotations

import argparse

from dxbridge_ctypes import (
    BACKEND_NAMES,
    DXB_FEATURE_DX12,
    DXBridgeLibrary,
    LOG_LEVEL_NAMES,
    parse_version,
)


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Load dxbridge.dll, print version, and receive logs."
    )
    parser.add_argument("--dll", help="Optional explicit path to dxbridge.dll")
    parser.add_argument(
        "--backend",
        choices=sorted(BACKEND_NAMES),
        default="auto",
        help="Backend hint passed to DXBridge_Init",
    )
    args = parser.parse_args()

    dxbridge = DXBridgeLibrary(args.dll)
    print(f"Loaded: {dxbridge.path}")

    def on_log(level: int, message: str, _user_data) -> None:
        name = LOG_LEVEL_NAMES.get(level, f"LEVEL_{level}")
        print(f"[log:{name}] {message}")

    dxbridge.set_log_callback(on_log)
    try:
        version = dxbridge.get_version()
        print(f"DXBridge version: {parse_version(version)} ({version:#010x})")

        dxbridge.init(BACKEND_NAMES[args.backend])
        supports_dx12 = bool(dxbridge.dll.DXBridge_SupportsFeature(DXB_FEATURE_DX12))
        print(f"DX12 feature flag reported by active backend: {supports_dx12}")
        print("Initialization and shutdown completed successfully.")
        dxbridge.shutdown()
    finally:
        dxbridge.set_log_callback(None)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
